File: common/scala_common/mcached_manipulate.scala
// --------------------------------------------------------------
//	mcached_manipulate.scala
//
//					Feb/03/2015
// --------------------------------------------------------------
import scala.io.Source
import scala.collection.mutable

import org.json4s._
import org.json4s.JsonDSL._
import org.json4s.native.JsonMethods._

import java.net.Socket
import java.io.DataInputStream
import java.io.InputStreamReader
import java.io.BufferedReader
import java.io.OutputStream
import	java.io.DataOutputStream
// --------------------------------------------------------------
object mcached_manipulate
{
// --------------------------------------------------------------
def get_record_proc (key_id:String,rr:InputStreamReader,output:OutputStream)
	:(mutable.Map[String,String]) = {
	var unit_aa = mutable.Map[String,String] ()

	val command = "get " + key_id + '\n'
	var data = command.getBytes()
	output.write (data)
	output.flush ()
	var key = ""
	var line:String = ""
	var buf = new BufferedReader(rr);

	var flag_find = false

	while(({line = buf.readLine(); line != null}) && line != "END")
		{
		if (line.contains ("VALUE"))
			{
			val cols = line.split (" ")
			key = cols(1)
			flag_find = true
			}
		else if (line.contains ("name"))
			{
	//		println (line)
			val json = parse(line)
			val name = (json \ "name").values
			val population = (json \ "population").values
			val date_mod = (json \ "date_mod").values
			unit_aa("name") = name.toString
			unit_aa("population") = population.toString
			unit_aa("date_mod") = date_mod.toString
			}
		else if (line.contains ("ERROR"))
			{
			println (line)
			return null
			}
		}

	if (flag_find == false)
		{
		unit_aa = null
		}

	unit_aa
}
// --------------------------------------------------------------
def mcached_to_dict_proc (server:String,port:Int,keys:Array[String]):(mutable.Map[String,Object]) = {

	var ss = new Socket (server,port)

	var input = ss.getInputStream ()
	var rr = new InputStreamReader (input)
	var output = ss.getOutputStream ()

	var dict_aa = mutable.Map[String,Object] ()

	for (key <- keys)
		{
		var unit_aa = get_record_proc (key,rr,output)

//		println ("key = " + key)

		if (unit_aa != null)
			{
			dict_aa(key) = unit_aa
			}
		}

	dict_aa
}

// --------------------------------------------------------------
def delete_record_proc (key_id:String,server:String,port:Int)
{
	println ("*** delete_record_proc aaa ***")

	var ss = new Socket (server,port)
	var input = ss.getInputStream ()
	var rr = new InputStreamReader (input)
	var output = ss.getOutputStream ()

	val command = "delete " + key_id + '\n'
	var data = command.getBytes()
	output.write (data)
	output.flush ()
	var key = ""
	var line = ""
	var buf = new BufferedReader(rr);
	println ("*** delete_record_proc ccc ***")
	while(({line = buf.readLine(); line != null}) && line != "END")
		{
			println (line)
		if (line.contains ("ERROR"))
			{
			println (line)
			return
			}
		if (line.contains ("NOT_FOUND"))
			{
			println (line)
			return
			}
		else if (line.contains ("DELETED"))
			{
			println (line)
			return
			}

		}
}

// --------------------------------------------------------------
def mcached_socket_write_proc (key:String ,value_in:String,
	input:DataInputStream,output:DataOutputStream)
{
	println (value_in)

	var rvalue = null
	val llx = value_in.getBytes().length
	val command = "set " + key + " 0 0 " + llx + "\r\n"
	println (command)

	output.writeBytes (command)
	output.flush ()

	var rr = new InputStreamReader (input)

	val command_bb =  value_in.getBytes()
	val rn = "\r\n"

	output.write (command_bb)
	output.writeBytes (rn)
	output.flush ()

	var buf = new BufferedReader(rr);

	var line:String = ""

	while(({line = buf.readLine(); line != null}) && line != "STORED")
		{
		if (line.contains ("ERROR"))
			{
			println (line)
//			return null
			}
		}

//	return	rvalue
}

// --------------------------------------------------------------
def dict_to_mcached_proc (server:String,port:Int,
	dict_aa: mutable.Map[String,Object])
{
	println ("*** dict_to_mcached_proc ***")

	var sock = new Socket (server,port)
	var output = new DataOutputStream (sock.getOutputStream ())
	var input = new DataInputStream (sock.getInputStream ())


	for (pair <- dict_aa)
		{
		val key = pair._1
		val unit_aa = pair._2.asInstanceOf [mutable.Map[String,String]]

		val unit_bb = unit_aa.toMap
		val json_str = compact(render(unit_bb))

		mcached_manipulate.mcached_socket_write_proc (key,json_str,input,output)
		}

	input.close ()
	output.close ()
	sock.close ()
}
// --------------------------------------------------------------
def mcached_update_proc (server:String,port:Int,key_in:String,population_in:Int)
{
	System.out.println ("*** mcached_update_proc ***");
	System.out.println ("key_in = " + key_in);

	var sock = new Socket (server,port)

	var iis = sock.getInputStream ()
	var input = new DataInputStream (iis)
	var rr = new InputStreamReader (iis)
	var output = new DataOutputStream (sock.getOutputStream ())

	var unit_aa = mcached_manipulate.get_record_proc (key_in,rr,output)

	if (unit_aa != null)
		{
		unit_aa("population") = "%d".format (population_in)
		val date_mod_new = text_manipulate.get_current_date_proc ()
		unit_aa("date_mod") = date_mod_new

		val unit_bb = unit_aa.toMap
		val json_str = compact(render(unit_bb))
		println (json_str)

		mcached_manipulate.mcached_socket_write_proc (key_in,json_str,input,output)
		}

	input.close ()
	output.close ()
	sock.close ()
}
// --------------------------------------------------------------
}

// --------------------------------------------------------------

File: common/rhino_common/mcached_manipulate.js
// ----------------------------------------------------------------------- 
/*
	mcached_manipulate.js

					Feb/05/2015

*/
// ----------------------------------------------------------------------- 
function socket_write_proc (key_in,value_in,is,os)
{
	var value_str = new java.lang.String (value_in);
	var llx = value_str.getBytes().length;

	var command = "set " + key_in + " 0 0 " + llx + "\r\n";
	os.writeBytes(command);

	var command_bb = value_str.getBytes();

	os.write (command_bb); 
	os.writeBytes ("\r\n"); 
	var rec_0 = is.readLine();
}

// ----------------------------------------------------------------------- 
function socket_read_proc (key_in,is,os)
{
	var json_str = "";

	os.writeBytes("get " + key_in + "\r\n"); 
	var rec_0 = is.readLine();

	if (rec_0 == "END")
		{
//		print (rec_0);
		}
	else
		{
		var json_str = is.readLine();
		rec_2 = is.readLine();
		}

	return	json_str;
}

// ----------------------------------------------------------------------- 
function mcached_fetch_proc (os,is,key_in)
{
	var json_str = socket_read_proc (key_in,is,os);

	if (json_str)
	{
	var unit_aa = JSON.parse (json_str);
	var name = unit_aa["name"];
	var population = unit_aa["population"];
	var date_mod = unit_aa["date_mod"];

	var out_str = key_in + "\t" + name + "\t";
	out_str += population + "\t" + date_mod;
	print (out_str);
	}
}

// ----------------------------------------------------------------------- 
function socket_delete_record_proc (out,is,key_in)
{
	out.writeBytes("delete " + key_in + "\r\n"); 
	var responseLine = is.readLine();
        print (responseLine);
}

// ----------------------------------------------------------------------- 
function mcached_to_dict_proc (server,port,keys)
{
	var dict_aa = new Object ()

	var socket = new Socket(server,port)
	var os = new DataOutputStream (socket.getOutputStream())
	var is = new BufferedReader
		(new InputStreamReader (socket.getInputStream()))

	if (socket != null && os != null && is != null)
		{
		for (var it in keys)
			{
			var key = keys[it]
			var json_str = socket_read_proc (key,is,os)
			if (json_str)
				{
				var unit_aa = JSON.parse (json_str)
				dict_aa[key] = unit_aa
				}

			}
		}

	os.close()
	is.close()
	socket.close()

	return	dict_aa
}

// ----------------------------------------------------------------------- 
function dict_to_mcached_proc (server,port,dict_aa)
{
	var socket = new Socket(server,port)
	var os = new DataOutputStream(socket.getOutputStream())
	var is = new BufferedReader
		(new InputStreamReader (socket.getInputStream())) 

	for (var key in dict_aa)
		{
		var str_json = JSON.stringify (dict_aa[key])
		print (str_json)

		socket_write_proc (key,str_json,is,os)
		}

	os.close()
	is.close()
	socket.close()
}

// ----------------------------------------------------------------------- 
function mcached_update_proc (server,port,key,population)
{
	var socket = new Socket (server,port)
	var os = new DataOutputStream(socket.getOutputStream())
	var is = new BufferedReader
		(new InputStreamReader (socket.getInputStream()))

	var json_str = socket_read_proc (key_in,is,os)

	print (json_str)

	if (json_str != null)
	{
	var unit_aa = JSON.parse (json_str)

	unit_aa['population'] = population_in
	unit_aa['date_mod'] = get_current_date_proc ()
	print ("key_in = " + key_in)

	var json_new = JSON.stringify (unit_aa)

	print (json_new)

	socket_write_proc (key_in,json_new,is,os)
	}

	os.close()
	is.close()
	socket.close() 
}

// ----------------------------------------------------------------------- 
function mcached_delete_proc (server,port,key_in)
{
	var socket = new Socket (server,port)
	var os = new DataOutputStream(socket.getOutputStream())
	var is = new BufferedReader
		(new InputStreamReader (socket.getInputStream()))

	if (socket != null && os != null && is != null)
		{
		socket_delete_record_proc (os,is,key_in)
		}

	os.close()
	is.close()
	socket.close()
}

// ----------------------------------------------------------------------- 
